File: ruby/inventory.rb
class InventoryItem < Struct.new :id, :name, :price
  def initialize id, name, price; super end
end

class StoreEvent < Struct.new :event_type, :payload, :timestamp
  def initialize event_type, payload, timestamp = Time.now.to_i; super end
end

class Store
  attr_accessor :name, :opening_time, :closing_time, :stock, :events

  def initialize name
    @name = name
    @opening_time = 900
    @closing_time = 1700
    @stock = {}
    @events = []
  end

  def update_item_quantity item, amount
    @stock[item.id] = amount
  end

  def purchase_item item
    if @stock[item.id] < 1
      return
    end

    @events.push StoreEvent.new :purchased, item.id
    @stock[item.id] = @stock[item.id] - 1
  end

  def refund_item item
    @events.push StoreEvent.new :refunded, item.id
    @stock[item.id] = @stock[item.id] + 1
  end
  
end

item1 = InventoryItem.new "1", "Item 1 - A", 500
item2 = InventoryItem.new "2", "Item 2 - B", 750
item3 = InventoryItem.new "3", "Item 3 - C", 1000

store = Store.new "Bob's Shop"

store.update_item_quantity item1, 10
store.update_item_quantity item2, 7
store.update_item_quantity item3, 5

store.purchase_item item1
store.refund_item item1

puts store.stock

File: elixir/inventory.exs
defmodule InventoryItem do
  defstruct [:id, :name, :price]
end

defmodule StoreEvent do
  defstruct [:event_type, :timestamp, :payload]
end

defmodule Store do
  defstruct [:name, :opening_time, :closing_time, :stock, :events]
end

defmodule StoreActions do
  def update_item_quantity(store, item, amount) do
    new_stock = Map.put(store.stock, item.id, amount)
    Map.merge(store, %{ stock: new_stock })
  end

  def purchase_item(store, item) do
    if Map.get(store.stock, item.id) > 0 do
      new_events = store.events ++ [%StoreEvent{ event_type: :purchased, timestamp: DateTime.utc_now, payload: item.id }]
      store = Map.merge(store, %{ events: new_events })
      StoreActions.update_item_quantity(store, item, Map.get(store.stock, item.id) - 1)
    else
      StoreActions.update_item_quantity(store, item, Map.get(store.stock, item.id))
    end
  end

  def refund_item(store, item) do
    new_events = store.events ++ [%StoreEvent{ event_type: :refunded, timestamp: DateTime.utc_now, payload: item.id }]
    store = Map.merge(store, %{ events: new_events })
    StoreActions.update_item_quantity(store, item, Map.get(store.stock, item.id) + 1)
  end
end

defmodule Main do
  def main do
    item1 = %InventoryItem{ id: 1, name: "Item 1 - A", price: 500 }
    item2 = %InventoryItem{ id: 2, name: "Item 2 - B", price: 750 }
    item3 = %InventoryItem{ id: 3, name: "Item 3 - C", price: 1000 }
    
    my_store = %Store{
      name: "Bob's Store",
      opening_time: nil,
      closing_time: nil,
      stock: %{},
      events: []
    }
    
    my_store = StoreActions.update_item_quantity(my_store, item1, 10)
    my_store = StoreActions.update_item_quantity(my_store, item2, 7)
    my_store = StoreActions.update_item_quantity(my_store, item3, 5)
    
    my_store = StoreActions.purchase_item(my_store, item1)
    my_store = StoreActions.purchase_item(my_store, item1)
    my_store = StoreActions.purchase_item(my_store, item2)
    my_store = StoreActions.purchase_item(my_store, item3)
    my_store = StoreActions.refund_item(my_store, item1)

    IO.inspect my_store
  end
end

Main.main
